Fix compilation issue kernels < 4.18 (Ubuntu 18.04 and derived distros)

CGarces opened the pull request:
As requested on #419 

Linux Mint 19 use a EOL kernel (like Ubuntu18.04)
I have made a new Travis build using similar kernel and gcc compiler.
 The fix for the issue just remove the function cfg80211_sinfo_alloc_tid_stats, introduced at https://github.com/torvalds/linux/commit/8689c051a20195b228e19acb155c7d6e48a86753#diff-35454f3f0395271a0e34581e160bea5f


Fix  #419 
Fix  #436

CGarces committed:
Travis CI: Create specific build for kernel 4.15.x (Ubuntu 18.04 and derived distros)

CGarces commented:
Commits squashed, ready to merge.

kimocoder commented:
Thanks a lot!

vcsomor commented:
Hi!

I've tried to compile (`sudo ./dkms-install.sh `) the code from @CGarces, on the following machine.
```
rtl8812au/CGares/rtl8812au$ uname -a
Linux hp-250-g5 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.3 LTS"
NAME="Ubuntu"
VERSION="18.04.3 LTS (Bionic Beaver)"
```

I'm getting this error by compiling `os_dep/linux/ioctl_cfg80211.c`

```
  CC [M]  /home/viktor/Downloads/rtl8812au/CGares/rtl8812au/os_dep/linux/ioctl_cfg80211.o
/home/viktor/Downloads/rtl8812au/CGares/rtl8812au/os_dep/linux/ioctl_cfg80211.c: In function ‘rtw_cfg80211_indicate_sta_assoc’:
/home/viktor/Downloads/rtl8812au/CGares/rtl8812au/os_dep/linux/ioctl_cfg80211.c:4484:3: error: implicit declaration of function ‘cfg80211_sinfo_alloc_tid_stats’; did you mean ‘cfg80211_inform_bss_width’? [-Werror=implicit-function-declaration]
   cfg80211_sinfo_alloc_tid_stats(&sinfo, GFP_KERNEL);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   cfg80211_inform_bss_width
/home/viktor/Downloads/rtl8812au/CGares/rtl8812au/os_dep/linux/ioctl_cfg80211.c: At top level:
cc1: warning: unrecognized command line option ‘-Wno-cast-function-type’
cc1: warning: unrecognized command line option ‘-Wno-cast-function-type’
cc1: some warnings being treated as errors
scripts/Makefile.build:330: recipe for target '/home/viktor/Downloads/rtl8812au/CGares/rtl8812au/os_dep/linux/ioctl_cfg80211.o' failed
make[2]: *** [/home/viktor/Downloads/rtl8812au/CGares/rtl8812au/os_dep/linux/ioctl_cfg80211.o] Error 1
Makefile:1571: recipe for target '_module_/home/viktor/Downloads/rtl8812au/CGares/rtl8812au' failed
make[1]: *** [_module_/home/viktor/Downloads/rtl8812au/CGares/rtl8812au] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-4.15.0-65-generic'
Makefile:1606: recipe for target 'modules' failed
make: *** [modules] Error 2
```

Git revision: `f6fd8a7c1abf6ed771b0f47019fe9f73673f96ee`

I see the [CI](https://travis-ci.org/aircrack-ng/rtl8812au/jobs/588193964) has the same error for the same kernel version.

CGarces commented:
Hi @vcsomor
The PR not fix the issue, only made it reproducible.
It include one TravisCI build for 4.15.x and GCC-7, adding  it to the matrix builds.

CGarces committed:
Fix commpilation issue for kernels < 4.18
cfg80211_sinfo_alloc_tid_stats function was included at 8689c051a20195b228e19acb155c7d6e48a86753

CGarces commented:
I have added and additional commit 19dcc3302edf4368064cab92007645348205ce75
@vcsomor , can you test it?
Also ping @vedantk and @iatsiuk who opened the issues.

iatsiuk commented:
I'm able to build module from commit https://github.com/aircrack-ng/rtl8812au/commit/19dcc3302edf4368064cab92007645348205ce75 but it crashes:

```
[ 1653.625029] usbcore: deregistering interface driver rtl88xxau
[ 1653.843001] usb 5-1: reset high-speed USB device number 3 using ehci-platform
[ 1667.871351] usbcore: registered new interface driver rtl88xxau
[ 1667.889488] rtl88xxau 5-1:1.0 wlxe84e0673ad97: renamed from wlan0
[ 1667.903307] Unable to handle kernel NULL pointer dereference at virtual address 000008c0
[ 1667.904142] pgd = ffffffc0e0a17000
[ 1667.904488] [000008c0] *pgd=0000000000000000, *pud=0000000000000000
[ 1667.905135] Internal error: Oops: 96000005 [#1] SMP
[ 1667.905626] Modules linked in: 88XXau(O) lzo zram midgard_kbase rockchip_saradc ip_tables x_tables autofs4 uas usb_storage [last unloaded: 88XXau]
[ 1667.907045] CPU: 4 PID: 3386 Comm: systemd-udevd Tainted: G           O    4.4.190-1233-rockchip-ayufan-gd3f1be0ed310 #1
[ 1667.908122] Hardware name: Pine64 RockPro64 (DT)
[ 1667.908587] task: ffffffc0df988000 task.stack: ffffffc0e09e0000
[ 1667.909478] PC is at rtw_ndev_notifier_call+0x54/0x70 [88XXau]
[ 1667.910349] LR is at rtw_ndev_notifier_call+0x30/0x70 [88XXau]
```

[dmesg.log](https://github.com/aircrack-ng/rtl8812au/files/3693404/dmesg.log)

CGarces commented:
Thanks @iatsiuk for your testing.
Looks like the error not is related with this patch see #409 #429
So in my opinion this PR is ready to merge.

CGarces commented:
@iatsiuk 
You are testing using my branch?
My PR branch is behind the upstream branch. The commit that fix the null pointer at rtw_ndev_notifier_call not exist 14 days ago when I created this PR, see 547d44bb9b9b45160530d4b75eb56f88018900e6
Make sure that you compile the code with the last version and apply one patch https://github.com/CGarces/rtl8812au/commit/19dcc3302edf4368064cab92007645348205ce75.patch

vedantk commented:
Hi @CGarces, thanks for taking a look at this. Regarding Issue #421, in my local testing, this does not appear to resolve the null pointer dereference in:

```
[  358.403506] BUG: kernel NULL pointer dereference, address: 00000000000008c0
...
[  358.403557]  notifier_call_chain+0x4c/0x70
[  358.403560]  dev_change_name+0x219/0x320
[  358.403562]  do_setlink+0xd45/0xf10
[  358.403565]  ? prep_new_page+0x46/0xd0
[  358.403567]  ? get_page_from_freelist+0xc03/0x1240
[  358.403568]  rtnl_setlink+0x105/0x170
[  358.403571]  rtnetlink_rcv_msg+0x137/0x3c0
[  358.403572]  ? rtnl_calcit.isra.0+0x120/0x120
[  358.403574]  netlink_rcv_skb+0x75/0x140
[  358.403576]  netlink_unicast+0x177/0x1f0
[  358.403578]  netlink_sendmsg+0x204/0x3d0
[  358.403580]  sock_sendmsg+0x5e/0x60
```

Here is what I tested:

```
~/src/rtl8812au (0) $ g log --oneline -n6
62b82be (HEAD -> v5.6.4.1) Merge remote-tracking branch 'CGarces/v5.6.4.1' into v5.6.4.1
19dcc33 (CGarces/v5.6.4.1) Fix commpilation issue for kernels < 4.18 cfg80211_sinfo_alloc_tid_stats function was included at 8689c051a20195b228e19acb155c7d6e48a86753
25c0645 (origin/v5.6.4.1, origin/HEAD) Merge pull request #426 from CGarces/patch-1
6b3af33 (CGarces/patch-1) Fix floating point in VHT_HW_TX for both x86 and rpi
547d44b Merge pull request #424 from u-ra/v5.6.4.1
4b162cb Fix oops on interface rename
```
